### DISCUSSION TITLE
Problem:Override time-axis tick-labels. Solution: Grafana Transform for the Timestamps themselves?

### DISCUSSION THREAD

**Discussion by samjewell** [grafana/grafana#56795] [Dashboards, Panels & Visualizations]
# Context

I'm building a dashboard to show customers what they consumed **each month**.

Specifically I need to query for the values that are present at the very end of each month, and show those, as that corresponds to end of the billing period. For the **September consumption** I would query at **midnight on the night of Sept 30th**. This actually corresponds to a datetime of `2022/10/01 00:00:00`.

I'm querying a Prometheus datasource, with this query:
```sql
sum(grafanacloud_logs_instance_bytes_received_per_second)
and day_of_month() == 1 and hour() == 0
```

# The problem

**This is unfortunate**, as it means that the tick-label rendered onto the panel (whether I choose Bar Chart or Time Series) is simply one of `2022-10` (seen on the Time-series panel) or `10/01` (seen on the Bar Chart panel). Remember we're trying to **show consumption for the month of September**.

**Bar Chart Panel**
<img src=https://user-images.githubusercontent.com/2903904/195335215-1a59574e-06a5-4653-9c53-d9d5569ce33d.png width=400>

**TimeSeries panel**
<img src=https://user-images.githubusercontent.com/2903904/195335414-f9356285-3003-4453-bca0-1f102812f4e3.png width=400>

At this point it is **misleading to the reader of that panel**: It looks like the bar is for October - ie for the consumption that happened during October. But this is wrong - the consumption rendered is actually for September.

# Desired solution

So I'd like a way to avoid misleading the viewer of the panel, and show September instead of October on the tick-label. I'd be happy with any implementation really, as long as I can achieve the end result.
One option which could be flexible and useful in other ways might be to:
- Add a Grafana Transform which would allow timestamps to be mutated / shortened. Basically I want to use a Grafana Transform to convert exactly `2022/10/01 00:00:00` into exactly `2022/09` only (ie. just the month, without any day or time component).

# A possible workaround

You might think I could query for the data at 1 second before midnight, in order to query within the same month as the consumption occurred. So I tried `2022/09/30 23:59:59` (or actually just `2022/09/30 23:59:00`) with the following query:

```
sum(grafanacloud_logs_instance_bytes_received_per_second)
and day_of_month() == days_in_month() and hour() == 23 and minute() == 59
```
But unfortunately I get "No data in response". I'm querying over a 1 year period, and the "Max data points" isn't high enough.
For a more successful workaround, I can use one hour before midnight - ie. `2022/09/30 23:00:00` **AND** dial up the "Max data points" to 8760. **NOTE** that this isn't ideal! The numbers I'm displaying are now 1 hour short of the total bill for the month, or about 0.1% out!
```
sum(grafanacloud_logs_instance_bytes_received_per_second)
and day_of_month() == days_in_month() and hour() == 23
```

This does not work for TimeSeries panel (although the tooltip you get when hovering does show the full timestamp: 
Not a fantastic screenshot here, but you should get the idea.
<img src=https://user-images.githubusercontent.com/2903904/195340040-cc8ff7e1-e7d5-4f0e-a6da-96a70a6b351d.png width=250>

It is only **partially successful** for the Bar Chart panel:
<img src=https://user-images.githubusercontent.com/2903904/195337586-8bb6ddd4-ef9a-445e-85a7-4eb187295f37.png width=400>

But it's now being impacted by Daylight Saving time for some reason! I'm based in the UK, using GMT (ie UTC) in the winter, and BST (UTC +1) in the summer time. The tick-marks from End-Oct thru End-Feb show how I want them to show, but the tick marks for the rest of the year still show month-start for the following month, rather than month-end of the preceding month:
<img src=https://user-images.githubusercontent.com/2903904/195337710-3b5db93e-becb-45e7-a274-ea4a35c0aa0e.png width=400

This itself feels like a separate bug: Why is it that when I use ` and day_of_month() == days_in_month() and hour() == 23` in my PromQL my own timezone is taken into account, even though the dashboard is clearly set to use UTC and not my own browser timezone or personal time.
<img src=https://user-images.githubusercontent.com/2903904/195338546-1cfdfc56-f9f4-4ae4-b8c2-6ed8612cf93c.png width=400>

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!

**Comment by samjewell** [comment]
If anyone is bothered, I found that the new "Format Time" transform helped me get to where I needed with this:
![image](https://github.com/grafana/grafana/assets/2903904/19a5eca9-39e6-4fe4-8313-34ed5da12b50)
